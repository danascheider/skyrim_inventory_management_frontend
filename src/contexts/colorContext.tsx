import { createContext, ReactElement } from 'react'
import { YELLOW, type ColorScheme } from '../utils/colorSchemes'

const ColorContext = createContext<ColorScheme>(YELLOW)

interface ProviderProps {
  colorScheme: ColorScheme
  children: ReactElement
}

const ColorProvider = ({ colorScheme, children }: ProviderProps) => (
  <ColorContext.Provider value={colorScheme}>{children}</ColorContext.Provider>
)

export { ColorProvider, ColorContext }
