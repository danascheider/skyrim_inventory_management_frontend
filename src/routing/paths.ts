export type PathString = `/${string}`

interface Paths {
  [i: string]: PathString
}

const paths: Paths = {
  home: '/',
  login: '/login'
}

export default paths
