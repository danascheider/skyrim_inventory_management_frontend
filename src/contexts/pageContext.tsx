import { createContext, useState, useEffect, useRef } from 'react'
import { type ApiCalls, type Resource, type HttpVerb } from '../types/apiCalls'
import { ProviderProps } from '../types/contexts'
import { type FlashProps, type ModalProps } from '../types/pageContext'

interface PageContextType {
  flashProps: FlashProps
  setFlashProps: (props: FlashProps) => void
  modalProps: ModalProps
  setModalProps: (props: ModalProps) => void
  apiCallsInProgress: ApiCalls
  addApiCall: (key: Resource, value: HttpVerb) => void
  removeApiCall: (key: Resource, value: HttpVerb) => void
}

const defaultFlashProps: FlashProps = {
  type: 'info',
  message: '',
  hidden: true,
}

const defaultModalProps: ModalProps = {
  hidden: true,
  children: <></>,
}

const defaultApiCallStatus = {
  games: [],
  shoppingLists: [],
  shoppingListItems: [],
}

const PageContext = createContext<PageContextType>({
  flashProps: defaultFlashProps,
  modalProps: defaultModalProps,
  apiCallsInProgress: defaultApiCallStatus,
  setFlashProps: () => {},
  setModalProps: () => {},
  addApiCall: () => {},
  removeApiCall: () => {},
})

const PageProvider = ({ children }: ProviderProps) => {
  const [flashProps, setFlashProps] = useState<FlashProps>(defaultFlashProps)
  const [modalProps, setModalProps] = useState<ModalProps>(defaultModalProps)
  const [apiCallsInProgress, SetApiCallsInProgress] =
    useState<ApiCalls>(defaultApiCallStatus)

  const flashVisibleSince = useRef(0)

  const addApiCall = (key: Resource, value: HttpVerb) => {
    SetApiCallsInProgress({
      ...apiCallsInProgress,
      [key]: [...apiCallsInProgress[key], value],
    })
  }

  const removeApiCall = (key: Resource, value: HttpVerb) => {
    SetApiCallsInProgress({
      ...apiCallsInProgress,
      [key]: apiCallsInProgress[key].filter((verb) => verb !== value),
    })
  }

  const value = {
    flashProps,
    modalProps,
    apiCallsInProgress,
    setFlashProps,
    setModalProps,
    addApiCall,
    removeApiCall,
  }

  useEffect(() => {
    if (flashProps.hidden === true) return

    flashVisibleSince.current = Number(new Date())

    setTimeout(() => {
      if (Number(new Date()) - flashVisibleSince.current >= 4000) {
        setFlashProps({ ...flashProps, hidden: true })
        flashVisibleSince.current = 0
      }
    }, 4000)
  }, [flashProps])

  return <PageContext.Provider value={value}>{children}</PageContext.Provider>
}

export { PageContext, PageProvider }
