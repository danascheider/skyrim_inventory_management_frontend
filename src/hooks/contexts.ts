import { useContext } from 'react'
import { ColorContext } from '../contexts/colorContext'
import { LoginContext } from '../contexts/loginContext'

const useCustomContext = <T>(cxt: React.Context<T>, msg: string) => {
  const context = useContext(cxt)

  if (!context) throw new Error(msg)

  return context
}

export const useColorScheme = () =>
  useCustomContext(
    ColorContext,
    'useColorScheme must be used within a ColorProvider'
  )

export const useGoogleLogin = () =>
  useCustomContext(
    LoginContext,
    'useGoogleLogin must be used within a LoginProvider'
  )