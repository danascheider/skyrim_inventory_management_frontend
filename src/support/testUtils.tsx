import { type ReactElement } from 'react'
import { BrowserRouter } from 'react-router-dom'
import { JSDOM } from 'jsdom'
import { render } from '@testing-library/react'
import { LoginContext } from '../contexts/loginContext'
import { User } from 'firebase/auth'
import testProfileImg from './testProfileImg.png'

export const BASE_APP_URI = 'http://localhost:5173'

declare global {
  namespace NodeJS {
    interface Global {
      document: Document
      window: Window
    }
  }
}

const setDom = () => {
  const dom = new JSDOM('<!doctype html><html><body></body></html>', {
    url: 'http://localhost:5173',
  })

  global.window = dom.window as unknown as Window & typeof globalThis
  global.document = dom.window.document
}

/*
 *
 * Global Test Data
 *
 */

export const requireLogin = () => {
  /* noop */
}

export const testUser = {
  uid: 'edna',
  displayName: 'Edna St. Vincent Millay',
  email: 'edna@gmail.com',
  photoURL: testProfileImg,
  emailVerified: true,
  getIdtoken: () =>
    new Promise<string>((resolve, _reject) => resolve('xxxxxxx')),
} as unknown as User

/*
 *
 * Test Renderers
 *
 */

const renderWithJSDOM = (ui: ReactElement) => {
  setDom()

  return render(ui)
}

export const renderWithRouter = (ui: ReactElement) =>
  renderWithJSDOM(<BrowserRouter>{ui}</BrowserRouter>)

export const renderAuthenticated = (ui: ReactElement, authLoading = false) =>
  renderWithRouter(
    <LoginContext.Provider
      value={{
        user: testUser,
        token: 'xxxxxxx',
        requireLogin,
        authLoading,
      }}
    >
      {ui}
    </LoginContext.Provider>
  )

export const renderUnauthenticated = (ui: ReactElement) =>
  renderWithRouter(
    <LoginContext.Provider
      value={{
        user: null,
        token: null,
        authLoading: false,
        requireLogin,
      }}
    >
      {ui}
    </LoginContext.Provider>
  )

export const renderAuthLoading = (ui: ReactElement) =>
  renderWithRouter(
    <LoginContext.Provider
      value={{ user: null, token: null, authLoading: true, requireLogin }}
    >
      {ui}
    </LoginContext.Provider>
  )
