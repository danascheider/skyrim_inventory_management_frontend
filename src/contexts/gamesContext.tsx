import { createContext, useCallback, useEffect, useState } from 'react'
import { signOutWithGoogle } from '../firebase'
import { type Game } from '../types/games'
import { type ProviderProps } from '../types/contexts'
import { useGoogleLogin, usePageContext } from '../hooks/contexts'
import { ApiError } from '../types/errors'
import { LOADING, DONE, ERROR, type LoadingState } from '../utils/loadingStates'
import { getGames, deleteGame } from '../utils/simApi'

// TODO: Add default values for context provider as follows:
//       const createGame = (game: Game) => {
//                                            postGames(game: Game, token: '')
//                                              .then((status: number, json: string) => /* noop */)
//                                          }
//       This will be useful in Storybook.

const NOT_FOUND_MESSAGE =
  "Oops! We couldn't find the game you're looking for. Please refresh and try again."
const UNEXPECTED_ERROR_MESSAGE =
  "Oops! Something unexpected went wrong. We're sorry! Please try again later."

export interface GamesContextType {
  games: Game[]
  gamesLoadingState: LoadingState
  // createGame: (game: Game) => void
  // updateGame: (gameId: number, attrs: Game) => void
  destroyGame: (gameId: number) => void
}

export const GamesContext = createContext<GamesContextType>({
  games: [],
  gamesLoadingState: LOADING,
  destroyGame: () => {},
})

export const GamesProvider = ({ children }: ProviderProps) => {
  const { user, token, authLoading, requireLogin } = useGoogleLogin()
  const [gamesLoadingState, setGamesLoadingState] = useState(LOADING)
  const [games, setGames] = useState<Game[]>([])
  const { setFlashProps } = usePageContext()

  /*
   *
   * Retrieve all the current user's games from the API
   *
   */

  const fetchGames = useCallback(() => {
    if (user && token) {
      getGames(token)
        .then(({ json }) => {
          if (Array.isArray(json)) {
            setGames(json)
            setGamesLoadingState(DONE)
          }
        })
        .catch((e: ApiError) => {
          if (import.meta.env.DEV)
            console.error(`Error ${e.name}: ${e.message}`)

          if (e.code === 401) {
            signOutWithGoogle()
          } else {
            setGames([])
            setGamesLoadingState(ERROR)
            setFlashProps({
              type: 'error',
              message: e.message,
              hidden: false,
            })
          }
        })
    }
  }, [user, token])

  /*
   *
   * Destroy the requested game and update the `games` array
   *
   */

  const destroyGame = useCallback(
    (gameId: number) => {
      if (user && token) {
        deleteGame(gameId, token)
          .then(({ status }) => {
            if (status === 204) {
              const newGames = games.filter(({ id }) => id !== gameId)
              setGames(newGames)
              setFlashProps({
                hidden: false,
                type: 'success',
                message: 'Success! Your game has been deleted.',
              })
            }
          })
          .catch((e: ApiError) => {
            if (import.meta.env.DEV)
              console.log(`Error ${e.name}: ${e.message}`)

            const message =
              e.code === 404 ? NOT_FOUND_MESSAGE : UNEXPECTED_ERROR_MESSAGE

            setFlashProps({
              hidden: false,
              type: 'error',
              message,
            })
          })
      }
    },
    [user, token, games]
  )

  const value = {
    games,
    gamesLoadingState,
    destroyGame,
  }

  useEffect(() => {
    requireLogin()
  }, [requireLogin])

  useEffect(() => {
    if (authLoading) return

    fetchGames()
  }, [authLoading, fetchGames])

  return <GamesContext.Provider value={value}>{children}</GamesContext.Provider>
}
