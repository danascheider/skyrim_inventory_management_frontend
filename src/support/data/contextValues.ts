import { type GamesContextType } from '../../contexts/gamesContext'
import { type LoginContextType } from '../../contexts/loginContext'
import { type ShoppingListsContextType } from '../../contexts/shoppingListsContext'
import { DONE, ERROR, LOADING } from '../../utils/loadingStates'
import { testUser } from './users'
import { allGames } from './games'
import { emptyShoppingLists, shoppingListsForGame } from './shoppingLists'

const noop = () => {}

/**
 *
 * Default values for Login context
 *
 */

export const loginContextValue: LoginContextType = {
  user: testUser,
  token: 'xxxxxxx',
  requireLogin: noop,
  authLoading: false,
}

export const loadingLoginContextValue: LoginContextType = {
  user: null,
  token: null,
  requireLogin: noop,
  authLoading: true,
}

export const unauthenticatedLoginContextValue: LoginContextType = {
  user: null,
  token: null,
  requireLogin: noop,
  authLoading: false,
}

/**
 *
 * Default value for Games context
 *
 */

export const gamesContextValue: GamesContextType = {
  games: allGames,
  gamesLoadingState: DONE,
  createGame: noop,
  updateGame: noop,
  destroyGame: noop,
}

export const gamesContextValueLoading: GamesContextType = {
  games: [],
  gamesLoadingState: LOADING,
  createGame: noop,
  updateGame: noop,
  destroyGame: noop,
}

export const gamesContextValueError: GamesContextType = {
  games: [],
  gamesLoadingState: ERROR,
  createGame: noop,
  updateGame: noop,
  destroyGame: noop,
}

/**
 *
 * Default values for Shopping Lists context
 *
 */

export const shoppingListsContextValueEmpty: ShoppingListsContextType = {
  shoppingLists: emptyShoppingLists,
  shoppingListsLoadingState: DONE,
  createShoppingList: noop,
}

export const shoppingListsContextValue: ShoppingListsContextType = {
  shoppingLists: shoppingListsForGame(77),
  shoppingListsLoadingState: DONE,
  createShoppingList: noop,
}

export const shoppingListsContextValueLoading: ShoppingListsContextType = {
  shoppingLists: emptyShoppingLists,
  shoppingListsLoadingState: LOADING,
  createShoppingList: noop,
}

export const shoppingListsContextValueError: ShoppingListsContextType = {
  shoppingLists: emptyShoppingLists,
  shoppingListsLoadingState: ERROR,
  createShoppingList: noop,
}
