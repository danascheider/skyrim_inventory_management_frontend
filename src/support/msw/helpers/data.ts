import {
  type RequestShoppingListItem,
  type RequestShoppingList,
  type ResponseShoppingListItem,
  type ResponseShoppingList,
} from '../../../types/apiData'
import { allGames } from '../../data/games'
import {
  allShoppingLists,
  shoppingListsForGame,
} from '../../data/shoppingLists'

const gameIds = allGames.map(({ id }) => id)

/**
 *
 * Shopping List Creation
 *
 */

export const newShoppingList = (
  attributes: RequestShoppingList,
  gameId: number
): ResponseShoppingList[] => {
  if (gameIds.indexOf(gameId) < 0)
    throw new Error(
      'Cannot generate shopping list for game that does not exist in test data'
    )

  const existingLists = shoppingListsForGame(gameId)

  if (!existingLists.length)
    throw new Error(
      'Cannot generate single list for game without existing aggregate'
    )

  return [
    {
      id: 93,
      game_id: gameId,
      aggregate: false,
      aggregate_list_id: existingLists[0].id,
      title: attributes.title || 'New Shopping List',
      list_items: [],
      created_at: new Date(),
      updated_at: new Date(),
    },
  ]
}

export const newShoppingListWithAggregate = (
  attributes: RequestShoppingList,
  gameId: number
): ResponseShoppingList[] => {
  if (gameIds.indexOf(gameId) < 0)
    throw new Error(
      'Cannot generate shopping list for game that does not exist in test data'
    )

  const existingLists = shoppingListsForGame(gameId)

  if (existingLists.length)
    throw new Error(
      'Cannot generate new aggregate list for game that already has one'
    )

  return [
    {
      id: 93,
      game_id: gameId,
      aggregate: true,
      aggregate_list_id: null,
      title: 'All Items',
      list_items: [],
      created_at: new Date(),
      updated_at: new Date(),
    },
    {
      id: 94,
      game_id: gameId,
      aggregate: false,
      aggregate_list_id: 93,
      title: attributes.title || 'My Shopping List 1',
      list_items: [],
      created_at: new Date(),
      updated_at: new Date(),
    },
  ]
}

/**
 *
 * Shopping list item creation
 *
 */

export const newShoppingListItem = (
  attributes: RequestShoppingListItem,
  listId: number
) => {
  const list = allShoppingLists.find(({ id }) => id === listId)

  if (!list)
    throw new Error(`No shopping list with ID ${listId} in the test data`)

  const shoppingList = { ...list }
  const allLists = shoppingListsForGame(shoppingList.game_id)
  const aggregateList = { ...allLists[0] }

  const newItem: ResponseShoppingListItem = {
    id: 42,
    list_id: listId,
    description: 'Dummy description for TypeScript',
    quantity: 1,
    unit_weight: null,
    notes: null,
    created_at: new Date(),
    updated_at: new Date(),
    ...attributes,
  }

  const newAggregateListItem: ResponseShoppingListItem = {
    ...newItem,
    id: 43,
    list_id: aggregateList.id,
  }

  aggregateList.list_items = [newAggregateListItem, ...aggregateList.list_items]
  shoppingList.list_items = [newItem, ...shoppingList.list_items]

  return [aggregateList, shoppingList]
}
