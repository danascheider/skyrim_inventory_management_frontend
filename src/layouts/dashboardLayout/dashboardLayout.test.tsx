import { describe, test, expect } from 'vitest'
import { setupServer } from 'msw/node'
import { waitFor } from '@testing-library/react'
import { BASE_APP_URI, renderAuthenticated } from '../../support/testUtils'
import { gamesContextValue } from '../../support/data/contextValues'
import {
  getGamesAllSuccess,
  getGamesEmptySuccess,
} from '../../support/msw/games'
import { GamesContext, GamesProvider } from '../../contexts/gamesContext'
import { PageProvider } from '../../contexts/pageContext'
import DashboardLayout from './dashboardLayout'
import paths from '../../routing/paths'

describe('<DashboardLayout>', () => {
  describe('when a title is given', () => {
    test('renders the title and content', () => {
      const wrapper = renderAuthenticated(
        <PageProvider>
          <GamesContext.Provider value={gamesContextValue}>
            <DashboardLayout title="Page Title">Hello World</DashboardLayout>
          </GamesContext.Provider>
        </PageProvider>
      )
      expect(wrapper).toBeTruthy()

      const h2 = wrapper.container.querySelector('h2')
      expect(h2?.textContent).toBe('Page Title')

      expect(wrapper.getByText('Hello World')).toBeTruthy()
    })

    test('renders the DashboardHeader', () => {
      const wrapper = renderAuthenticated(
        <PageProvider>
          <GamesContext.Provider value={gamesContextValue}>
            <DashboardLayout title="Page Title">Hello World</DashboardLayout>
          </GamesContext.Provider>
        </PageProvider>
      )

      const a = wrapper.container.querySelector('a')
      expect(a?.textContent).toBe('Skyrim Inventory Management')
      expect(a?.href).toBe(`${BASE_APP_URI}${paths.dashboard.main}`)

      expect(wrapper.getByText('Edna St. Vincent Millay')).toBeTruthy()
      expect(wrapper.getByText('edna@gmail.com')).toBeTruthy()
    })

    test("doesn't display the StyledSelect", () => {
      const wrapper = renderAuthenticated(
        <PageProvider>
          <GamesContext.Provider value={gamesContextValue}>
            <DashboardLayout title="Page Title">Hello World</DashboardLayout>
          </GamesContext.Provider>
        </PageProvider>
      )

      expect(wrapper.queryByTestId('styledSelect')).toBeFalsy()
    })

    test('matches snapshot', () => {
      const wrapper = renderAuthenticated(
        <PageProvider>
          <GamesContext.Provider value={gamesContextValue}>
            <DashboardLayout title="Page Title">Hello World</DashboardLayout>
          </GamesContext.Provider>
        </PageProvider>
      )

      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('when no title is given', () => {
    test('displays content but not an h2 or hr', () => {
      const wrapper = renderAuthenticated(
        <PageProvider>
          <GamesContext.Provider value={gamesContextValue}>
            <DashboardLayout>Hello World</DashboardLayout>
          </GamesContext.Provider>
        </PageProvider>
      )
      expect(wrapper).toBeTruthy()

      const h2 = wrapper.container.querySelector('h2')
      expect(h2).toBeFalsy()

      const hr = wrapper.container.querySelector('hr')
      expect(hr).toBeFalsy()

      expect(wrapper.getByText('Hello World')).toBeTruthy()
    })

    test('renders the DashboardHeader', () => {
      const wrapper = renderAuthenticated(
        <PageProvider>
          <GamesContext.Provider value={gamesContextValue}>
            <DashboardLayout>Hello World</DashboardLayout>
          </GamesContext.Provider>
        </PageProvider>
      )

      const a = wrapper.container.querySelector('a')
      expect(a?.textContent).toBe('Skyrim Inventory Management')
      expect(a?.href).toBe(`${BASE_APP_URI}${paths.dashboard.main}`)

      expect(wrapper.getByText('Edna St. Vincent Millay')).toBeTruthy()
      expect(wrapper.getByText('edna@gmail.com')).toBeTruthy()
    })

    test("doesn't render the StyledSelect", () => {
      const wrapper = renderAuthenticated(
        <PageProvider>
          <GamesContext.Provider value={gamesContextValue}>
            <DashboardLayout>Hello World</DashboardLayout>
          </GamesContext.Provider>
        </PageProvider>
      )

      expect(wrapper.queryByTestId('styledSelect')).toBeFalsy()
    })

    test('matches snapshot', () => {
      const wrapper = renderAuthenticated(
        <PageProvider>
          <GamesContext.Provider value={gamesContextValue}>
            <DashboardLayout>Hello World</DashboardLayout>
          </GamesContext.Provider>
        </PageProvider>
      )

      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('when includeGameSelector is set to true', () => {
    describe('when games are returned from the API', () => {
      const mockServer = setupServer(getGamesAllSuccess)

      beforeAll(() => mockServer.listen())
      beforeEach(() => mockServer.resetHandlers())
      afterAll(() => mockServer.close())

      test('renders the select box', () => {
        const wrapper = renderAuthenticated(
          <PageProvider>
            <GamesProvider>
              <DashboardLayout title="Your Games" includeGameSelector>
                Hello World
              </DashboardLayout>
            </GamesProvider>
          </PageProvider>
        )

        expect(wrapper).toBeTruthy()
        expect(wrapper.getByTestId('styledSelect')).toBeTruthy()
      })

      test('includes all the games on the list', async () => {
        const wrapper = renderAuthenticated(
          <PageProvider>
            <GamesProvider>
              <DashboardLayout title="Your Games" includeGameSelector>
                Hello World
              </DashboardLayout>
            </GamesProvider>
          </PageProvider>
        )

        const selectedOption = wrapper.getByTestId('selectedOption')

        // Initial loading value
        expect(selectedOption.textContent).toEqual('Games loading...')

        await waitFor(() => {
          expect(selectedOption.textContent).toEqual('')
          expect(wrapper.getByText('My Game 1')).toBeTruthy()
          expect(wrapper.getByText('My Game 2')).toBeTruthy()
          expect(wrapper.getByText('Game with a really real...')).toBeTruthy()
        })
      })

      test('matches snapshot', () => {
        const wrapper = renderAuthenticated(
          <PageProvider>
            <GamesProvider>
              <DashboardLayout title="Your Games" includeGameSelector>
                Hello World
              </DashboardLayout>
            </GamesProvider>
          </PageProvider>
        )

        expect(wrapper).toMatchSnapshot()
      })
    })

    describe('when there are no games available', () => {
      const mockServer = setupServer(getGamesEmptySuccess)

      beforeAll(() => mockServer.listen())
      beforeEach(() => mockServer.resetHandlers())
      afterAll(() => mockServer.close())

      test('displays a placeholder', async () => {
        const wrapper = renderAuthenticated(
          <PageProvider>
            <GamesProvider>
              <DashboardLayout includeGameSelector>Hello World</DashboardLayout>
            </GamesProvider>
          </PageProvider>
        )

        const selectedOption = wrapper.getByTestId('selectedOption')
        expect(selectedOption.textContent).toEqual('Games loading...')

        await waitFor(() => {
          expect(selectedOption.textContent).toEqual('No games available')
        })
      })

      test('matches snapshot', () => {
        const wrapper = renderAuthenticated(
          <PageProvider>
            <GamesProvider>
              <DashboardLayout includeGameSelector>Hello World</DashboardLayout>
            </GamesProvider>
          </PageProvider>
        )

        expect(wrapper).toMatchSnapshot()
      })
    })

    describe('when a game is selected in the query string', () => {
      describe('when the selected game corresponds to a game in the games array', () => {
        const mockServer = setupServer(getGamesAllSuccess)

        beforeAll(() => mockServer.listen())
        beforeEach(() => mockServer.resetHandlers())
        afterAll(() => mockServer.close())

        test('sets the selected game as the default option', async () => {
          const wrapper = renderAuthenticated(
            <PageProvider>
              <GamesProvider>
                <DashboardLayout includeGameSelector>
                  Hello World
                </DashboardLayout>
              </GamesProvider>
            </PageProvider>,
            'http://localhost:5173/shoppingLists?gameId=51'
          )

          const selectedOption = wrapper.getByTestId('selectedOption')
          expect(selectedOption.textContent).toEqual('Games loading...')

          await waitFor(() => {
            expect(selectedOption.textContent).toEqual('My Game 2')
          })
        })

        test('matches snapshot', () => {
          const wrapper = renderAuthenticated(
            <PageProvider>
              <GamesContext.Provider value={gamesContextValue}>
                <DashboardLayout includeGameSelector>
                  Hello World
                </DashboardLayout>
              </GamesContext.Provider>
            </PageProvider>,
            'http://localhost:5173/shoppingLists?gameId=51'
          )

          expect(wrapper).toMatchSnapshot()
        })
      })

      describe('when the selected game does not correspond to a game in the games array', () => {
        const mockServer = setupServer(getGamesAllSuccess)

        beforeAll(() => mockServer.listen())
        beforeEach(() => mockServer.resetHandlers())
        afterAll(() => mockServer.close())

        test('sets the selected game as the default option', async () => {
          const wrapper = renderAuthenticated(
            <PageProvider>
              <GamesProvider>
                <DashboardLayout includeGameSelector>
                  Hello World
                </DashboardLayout>
              </GamesProvider>
            </PageProvider>,
            'http://localhost:5173/shoppingLists?gameId=67'
          )

          const selectedOption = wrapper.getByTestId('selectedOption')
          expect(selectedOption.textContent).toEqual('Games loading...')

          await waitFor(() => {
            expect(selectedOption.textContent).toEqual('')
          })
        })

        test('matches snapshot', () => {
          const wrapper = renderAuthenticated(
            <PageProvider>
              <GamesContext.Provider value={gamesContextValue}>
                <DashboardLayout includeGameSelector>
                  Hello World
                </DashboardLayout>
              </GamesContext.Provider>
            </PageProvider>,
            'http://localhost:5173/shoppingLists?gameId=67'
          )

          expect(wrapper).toMatchSnapshot()
        })
      })
    })
  })
})
