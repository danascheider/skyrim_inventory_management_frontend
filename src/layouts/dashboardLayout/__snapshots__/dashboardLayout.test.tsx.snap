// Vitest Snapshot v1

exports[`<DashboardLayout> > when a title is given > matches snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <main
        class="_root_03d657"
      >
        <section
          class="_container_03d657"
        >
          <h2
            class="_title_03d657"
          >
            Page Title
          </h2>
          <hr
            class="_hr_03d657"
          />
          Hello World
        </section>
        <header
          class="_root_c29812"
        >
          <div
            class="_bar_c29812"
          >
            <span
              class="_headerContainer_c29812"
            >
              <h1
                class="_header_c29812"
              >
                <a
                  class="_headerLinkLarge_c29812"
                  href="/dashboard"
                >
                  Skyrim Inventory
                  <br
                    class="_bp_c29812"
                  />
                   Management
                </a>
                <a
                  class="_headerLinkSmall_c29812"
                  href="/dashboard"
                >
                  S. I. M.
                </a>
              </h1>
            </span>
          </div>
        </header>
      </main>
    </div>
  </body>,
  "container": <div>
    <main
      class="_root_03d657"
    >
      <section
        class="_container_03d657"
      >
        <h2
          class="_title_03d657"
        >
          Page Title
        </h2>
        <hr
          class="_hr_03d657"
        />
        Hello World
      </section>
      <header
        class="_root_c29812"
      >
        <div
          class="_bar_c29812"
        >
          <span
            class="_headerContainer_c29812"
          >
            <h1
              class="_header_c29812"
            >
              <a
                class="_headerLinkLarge_c29812"
                href="/dashboard"
              >
                Skyrim Inventory
                <br
                  class="_bp_c29812"
                />
                 Management
              </a>
              <a
                class="_headerLinkSmall_c29812"
                href="/dashboard"
              >
                S. I. M.
              </a>
            </h1>
          </span>
        </div>
      </header>
    </main>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`<DashboardLayout> > when no title is given > matches snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <main
        class="_root_03d657"
      >
        <section
          class="_container_03d657"
        >
          Hello World
        </section>
        <header
          class="_root_c29812"
        >
          <div
            class="_bar_c29812"
          >
            <span
              class="_headerContainer_c29812"
            >
              <h1
                class="_header_c29812"
              >
                <a
                  class="_headerLinkLarge_c29812"
                  href="/dashboard"
                >
                  Skyrim Inventory
                  <br
                    class="_bp_c29812"
                  />
                   Management
                </a>
                <a
                  class="_headerLinkSmall_c29812"
                  href="/dashboard"
                >
                  S. I. M.
                </a>
              </h1>
            </span>
          </div>
        </header>
      </main>
    </div>
  </body>,
  "container": <div>
    <main
      class="_root_03d657"
    >
      <section
        class="_container_03d657"
      >
        Hello World
      </section>
      <header
        class="_root_c29812"
      >
        <div
          class="_bar_c29812"
        >
          <span
            class="_headerContainer_c29812"
          >
            <h1
              class="_header_c29812"
            >
              <a
                class="_headerLinkLarge_c29812"
                href="/dashboard"
              >
                Skyrim Inventory
                <br
                  class="_bp_c29812"
                />
                 Management
              </a>
              <a
                class="_headerLinkSmall_c29812"
                href="/dashboard"
              >
                S. I. M.
              </a>
            </h1>
          </span>
        </div>
      </header>
    </main>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
